executor/write.go

```diff
@@ -846,13 +844,13 @@ func (e *InsertExec) exec(ctx context.Context, rows [][]types.Datum) (types.Datu
 
 	// If `ON DUPLICATE KEY UPDATE` is specified, and no `IGNORE` keyword,
 	// the to-be-insert rows will be check on duplicate keys and update to the new rows.
-	if len(e.OnDuplicate) > 0 && !e.IgnoreErr {
+	if len(e.OnDuplicate) > 0 && !e.ctx.GetSessionVars().StmtCtx.IgnoreErr {
```

Comment:
Extract `e.ctx.GetSessionVars().StmtCtx.IgnoreErr` as `ignoreErr` outside this if branch,
since it can be reused.